lib/semantics/check-allocate.cc

```diff
@@ -0,0 +1,411 @@
+    return true;
+  }
+  if (const IntrinsicTypeSpec * intrinsicType2{type2.AsIntrinsic()}) {
+    if (const IntrinsicTypeSpec * intrinsicType1{type1.AsIntrinsic()}) {
+      return intrinsicType1->category() == intrinsicType2->category();
+    } else {
+      return false;
+    }
+  } else if (const DerivedTypeSpec * derivedType2{type2.AsDerived()}) {
+    return IsTypeCompatible(type1, *derivedType2);
+  }
+  return false;
+}
+
+static bool IsTypeCompatible(
+    const DeclTypeSpec &type1, const evaluate::DynamicType &type2) {
+  if (type1.category() == DeclTypeSpec::Category::ClassStar) {
+    // TypeStar does not make sens in allocate context because assumed type
+    // cannot be allocatable (C709)
+    return true;
+  }
+  if (type2.category != evaluate::TypeCategory::Derived) {
+    if (const IntrinsicTypeSpec * intrinsicType1{type1.AsIntrinsic()}) {
+      return intrinsicType1->category() == type2.category;
+    } else {
+      return false;
+    }
+  } else {
+    CHECK(type2.derived);
+    return IsTypeCompatible(type1, *type2.derived);
+  }
+  return false;
+}
+
+// Note: Check assumes  type1 is compatible with type2. type2 may have more type
+// parameters than type1 but if a type2 type parameter is assumed, then this
+// check enforce that type1 has it. type1 can be unlimited polymorphic, but not
+// type2.
+static bool HaveSameAssumedTypeParameters(
+    const DeclTypeSpec &type1, const DeclTypeSpec &type2) {
+  if (type2.category() == DeclTypeSpec::Category::Character) {
+    bool type2LengthIsAssumed{type2.characterTypeSpec().length().isAssumed()};
+    if (type1.category() == DeclTypeSpec::Category::Character) {
+      return type1.characterTypeSpec().length().isAssumed() ==
+          type2LengthIsAssumed;
+    }
+    // It is possible to reach this if type1 is unlimited polymorphic
+    return !type2LengthIsAssumed;
+  } else if (const DerivedTypeSpec * derivedType2{type2.AsDerived()}) {
+    int type2AssumedParametersCount{0};
+    int type1AssumedParametersCount{0};
+    for (auto it{derivedType2->parameters().begin()};
+         it != derivedType2->parameters().end(); ++it) {
+      type2AssumedParametersCount += it->second.isAssumed();
+    }
+    if (const DerivedTypeSpec *
+        derivedType1{
+            type1.AsDerived()}) {  // type1 may be unlimited polymorphic
+      for (auto it{derivedType1->parameters().begin()};
+           it != derivedType1->parameters().end(); ++it) {
+        if (it->second.isAssumed()) {
+          ++type1AssumedParametersCount;
+          if (const ParamValue *
+              param{derivedType2->FindParameter(it->first)}) {
+            if (!param->isAssumed()) {
+              return false;  // type1 has an assumed param that is not assumed
+                             // in type2
+            }
+          } else {
+            return false;  // type1 has an assumed param that is not a type
+                           // param of type2.
+          }
```

Comment:
Consider writing this if statement like this:
```
const ParamValue *param{derivedType2->FindParameter(it->first)};
if (!param || !param->isAssumed()) {
  return false;
}
```

Reply:
Applied